File: 241011/새치기 2/cut-in-line2.py
from collections import deque

# 입력 처리
N, M, Q = map(int, input().split())
people = input().split()

# M개의 줄을 만듦 (각 줄은 deque로 처리)
lines = [deque() for _ in range(M)]

# 각 사람을 해당 줄에 배치
for i, person in enumerate(people):
    line_idx = i % M
    lines[line_idx].append(person)

# 사람의 이름을 인덱스 찾기 쉽게 딕셔너리로 만듦
position = {person: i % M for i, person in enumerate(people)}

# 명령 처리
for _ in range(Q):
    command = input().split()
    
    if command[0] == '1':
        # 1 a b : a가 b 앞에 새치기
        a, b = command[1], command[2]
        line_idx = position[a]
        lines[line_idx].remove(a)  # a를 원래 위치에서 제거
        idx_b = lines[line_idx].index(b)  # b의 위치를 찾음
        lines[line_idx].insert(idx_b, a)  # b 앞에 a 삽입
    
    elif command[0] == '2':
        # 2 a : a가 집으로 감
        a = command[1]
        line_idx = position[a]
        lines[line_idx].remove(a)  # a를 줄에서 제거
    
    elif command[0] == '3':
        # 3 a b c : a부터 b까지 사람들을 떼어내어 c 앞으로 이동
        a, b, c = command[1], command[2], command[3]
        line_a_b = position[a]  # a와 b가 서 있는 줄
        line_c = position[c]    # c가 서 있는 줄
        
        # a부터 b까지의 사람들을 찾음
        idx_a = lines[line_a_b].index(a)
        idx_b = lines[line_a_b].index(b)
        group = deque()  # 떼어낼 그룹을 저장할 덱
        
        for i in range(idx_a, idx_b + 1):
            group.append(lines[line_a_b][i])
        
        # 원래 줄에서 a부터 b까지 제거
        for i in range(idx_b, idx_a - 1, -1):
            lines[line_a_b].remove(lines[line_a_b][i])
        
        # c 앞에 그룹을 삽입
        idx_c = lines[line_c].index(c)
        for person in reversed(group):
            lines[line_c].insert(idx_c, person)

# 결과 출력
for line in lines:
    if line:
        print(' '.join(line))
    else:
        print(-1)

File: 241011/새치기 2/cut-in-line2.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>

using namespace std;

int main() {
    int N, M, Q;
    cin >> N >> M >> Q;

    vector<deque<string>> lines(M); // 각 줄을 관리할 deque
    unordered_map<string, int> person_line; // 사람 이름과 그가 속한 줄을 매핑

    vector<string> people(N);
    for (int i = 0; i < N; ++i) {
        cin >> people[i];
        int line_idx = i / (N / M); // 각 사람을 적절한 줄에 분배
        lines[line_idx].push_back(people[i]);
        person_line[people[i]] = line_idx;
    }

    for (int i = 0; i < Q; ++i) {
        int type;
        cin >> type;

        if (type == 1) {
            string a, b;
            cin >> a >> b;
            int line_a = person_line[a];
            int line_b = person_line[b];

            if (line_a != line_b) {
                // a를 제거하고 b 앞에 삽입
                for (auto it = lines[line_a].begin(); it != lines[line_a].end(); ++it) {
                    if (*it == a) {
                        lines[line_a].erase(it);
                        break;
                    }
                }
                for (auto it = lines[line_b].begin(); it != lines[line_b].end(); ++it) {
                    if (*it == b) {
                        lines[line_b].insert(it, a);
                        break;
                    }
                }
                person_line[a] = line_b;
            }
        } else if (type == 2) {
            string a;
            cin >> a;
            int line_a = person_line[a];

            // a를 줄에서 제거
            for (auto it = lines[line_a].begin(); it != lines[line_a].end(); ++it) {
                if (*it == a) {
                    lines[line_a].erase(it);
                    break;
                }
            }
            person_line.erase(a); // person_line 에서도 제거
        } else if (type == 3) {
            string a, b, c;
            cin >> a >> b >> c;
            int line_a = person_line[a];
            int line_c = person_line[c];

            // a부터 b까지 통째로 c 앞에 이동
            deque<string> moving_people;
            bool collect = false;
            for (auto it = lines[line_a].begin(); it != lines[line_a].end();) {
                if (*it == a) collect = true;
                if (collect) {
                    moving_people.push_back(*it);
                    it = lines[line_a].erase(it);
                    if (*it == b) break;
                } else {
                    ++it;
                }
            }

            for (auto it = lines[line_c].begin(); it != lines[line_c].end(); ++it) {
                if (*it == c) {
                    for (auto &person : moving_people) {
                        lines[line_c].insert(it, person);
                        person_line[person] = line_c;
                    }
                    break;
                }
            }
        }
    }

    // 결과 출력
    for (int i = 0; i < M; ++i) {
        if (lines[i].empty()) {
            cout << "-1" << endl;
        } else {
            for (auto &person : lines[i]) {
                cout << person << " ";
            }
            cout << endl;
        }
    }

    return 0;
}